Editors/RoomEditor.cpp

```diff
@@ -1,13 +1,141 @@
 #include "RoomEditor.h"
+#include "Components/ArtManager.h"
+#include "Components/QMenuView.h"
+#include "MainWindow.h"
 #include "ui_RoomEditor.h"
 
+#include "Models/ImmediateMapper.h"
+#include "Models/TreeSortFilterProxyModel.h"
+#include "codegen/Room.pb.h"
+
+#include <QGraphicsPixmapItem>
 #include <QGraphicsScene>
+#include <QMouseEvent>
+
+#include <algorithm>
+
+using View = buffers::resources::Room::View;
 
 RoomEditor::RoomEditor(ProtoModel* model, QWidget* parent) : BaseEditor(model, parent), ui(new Ui::RoomEditor) {
   ui->setupUi(this);
 
-  QGraphicsScene* scene = new QGraphicsScene(this);
-  ui->graphicsView->setScene(scene);
+  ProtoModel* roomModel = model->GetSubModel(TreeNode::kRoomFieldNumber);
+  ui->roomRenderer->SetResourceModel(roomModel);
+
+  nodeMapper->addMapping(ui->roomName, TreeNode::kNameFieldNumber);
+  nodeMapper->toFirst();
+
+  resMapper->addMapping(ui->speedSpinBox, Room::kSpeedFieldNumber);

```

Comment:
This is the cruft I was suggesting getting rid of using attributes or properties or whatever on the Qt XML. I think you mentioned some snag you hit doing that in Discord, but I don't remember it. If it was just that these sources are precompiled, that doesn't actually matter. If you just put a `protofield` attribute in the UI elements, you could automatically look them up in a generalized loop.

Reply:
Yes, I said we could maybe do that using dynamic properties. But this is more complicated now that we have begun mapping nested fields like with rooms.
http://doc.qt.io/archives/qt-4.8/designer-widget-mode.html#dynamic-properties

Let me say first it's very obvious we could do something like string "views/visible" path properties to the protofield of each widget. But I hate this for two reasons, we would have to recurse over the widget hierarchy to look up this dynamic property to make the mapping. Also, it's not quite as fail safe as our mapper code which is checked at compile time (though warnings on the dynamic mapping would make it easier to detect an incorrect mapping).

The problem with QDataWidgetMapper and our own derived ImmediateMapper is that they are mainly oriented at master-detail interfaces. Hence why we are having to create new mappers for each submodel/subproto message.

Reply:
It's not going to get easier, and it's not going to be a one-size fits all solution. If you can do half the IDE like that, you'll save *thousands* of lines of useless boilerplate code. Had I done this project on my own, I'd probably have used wx and written my own YAML-based UI parser with these sorts of features. Qt gives us something workable out of the box. Use it where you can.